internal/expr/prepare.go

```diff
@@ -26,21 +36,145 @@ func getKeys(m map[string]*info) []string {
 	return keys
 }
 
-// prepareInput checks that the input expression corresponds to a known type.
-func prepareInput(ti typeNameToInfo, p *inputPart) error {
-	info, ok := ti[p.source.prefix]
-	if !ok {
-		return fmt.Errorf(`type %s unknown, have: %s`, p.source.prefix, strings.Join(getKeys(ti), ", "))
-	}
+// starCheckInput checks that the input is well formed with regard to
+// asterisks and the number of sources and targets.
+func starCheckInput(p *inputPart) error {
+	numColumns := len(p.cols)
+	numSources := len(p.source)
 
-	if _, ok = info.tagToField[p.source.name]; !ok {
-		return fmt.Errorf(`type %s has no %q db tag`,
-			info.structType.Name(), p.source.name)
+	sourceStars := starCount(p.source)
+	columnStars := 0
+	for _, col := range p.cols {
+		if col == "*" {
+			columnStars++
+		}
 	}
+	starSource := sourceStars == 1
+	starColumn := columnStars == 1
 
+	if sourceStars > 1 || columnStars > 1 || (columnStars == 1 && sourceStars == 0) ||
+		(starColumn && numColumns > 1) || (starSource && numSources > 1) || (starSource && numColumns == 0) {
+		return fmt.Errorf("invalid asterisk in input expression: %s", p)
+	}
+	if !starSource && (numColumns > 0 && (numColumns != numSources)) {
+		return fmt.Errorf("mismatched number of inputs and cols in input expression: %s", p)
+	}
 	return nil
 }
 
+func starCount(fns []fullName) int {
+	s := 0
+	for _, fn := range fns {
+		if fn.name == "*" {
+			s++
+		}
+	}
+	return s
+}
+
+func printList(xs []string) string {
+	var s bytes.Buffer
+	s.WriteString("(")
+	s.WriteString(strings.Join(xs, ", "))
+	s.WriteString(")")
+	return s.String()
+}
+
+func nParams(start int, num int) string {
+	var s bytes.Buffer
+	s.WriteString("(")
+	for i := start; i < start+num; i++ {
+		s.WriteString("@_sqlair_")
+		s.WriteString(strconv.Itoa(i))
+		if i < start+num-1 {
+			s.WriteString(", ")
+		}
+	}
+	s.WriteString(")")
+	return s.String()
+}
+
+// prepareInput first checks the types mentioned in the expression are known, it
+// then checks the expression is valid and generates the SQL to print in place
+// of it.
+// As well as the SQL string it also returns a list of fullNames containing the
+// type and tag of each input parameter. These are used in the complete stage to
+// extract the arguments from the relevent structs.
+func prepareInput(ti typeNameToInfo, p *inputPart, n int) (string, []inputLocation, error) {
```

Comment:
I think this doc exposes a few implementation details, perhaps it could be simplified to something like this?
```go
// prepareInput generates SQL identifiers and associated input locations for the given input part.
```

Reply:
True, the comment is also incorrect haha so worth changing anyway